### PR TITLE
[libc] Add proxy header math_macros.h.

### PR DESCRIPTION
**Pull request by lntue**
Context: https://github.com/llvm/llvm-project/pull/87017

- Add proxy header `libc/hdr/math_macros.h` that will:
  - include `<math.h>` in overlay mode,
  - include `"include/llvm-libc-macros/math-macros.h"` in full build mode.
- Its corresponding CMake target `libc.hdr.math_macros` will only depend on `libc.include.math` and `libc.include.llvm-libc-macros.math_macros` in full build mode.
- Replace all `#include "include/llvm-libc-macros/math-macros.h"` with `#include "hdr/math_macros.h"`.
- Add dependency to `libc.hdr.math_macros` CMake target when using `add_fp_unittest`.
- Update the remaining dependency.
- Update bazel overlay: add `libc:hdr_math_macros` target, and replacing all dependency on `libc:llvm_libc_macros_math_macros` with `libc:hdr_math_macros`.

### CODE REVIEW

**Review by SchrodingerZhu — COMMENTED** [review]
The patch overall looks good to me. Do we expose any macro to end users to indicate full build and overlay build?

**Comment by lntue** [comment]
> The patch overall looks good to me. Do we expose any macro to end users to indicate full build and overlay build?

For cmake cli, `-DLLVM_LIBC_FULL_BUILD=ON` should do it.  For user codes:
- Once everything is routed through proxy headers, there won't be any `include/*` headers, and users will just include system headers.
- For full build mode, since our `include/*` headers will now only be included in this mode, and they can assume that we are the only headers of the system, and don't need to do any of the `#include_next` dances.

For implementations in `src` folder, we can use `#ifdef LIBC_FULL_BUILD` to check whether we are in full build mode or overlay mode.

**Review by SchrodingerZhu — APPROVED** [review]
LGTM

File: libc/hdr/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  if(LLVM_LIBC_FULL_BUILD AND ADD_PROXY_HEADER_FULL_BUILD_DEPENDS)
+    list(APPEND deps ${ADD_PROXY_HEADER_FULL_BUILD_DEPENDS})
+  endif()
+
+  add_header_library(
+    ${target_name}
+    ${ADD_PROXY_HEADER_UNPARSED_ARGUMENTS}
+    DEPENDS ${deps}
+  )
+endfunction()
```
Comment (nickdesaulniers) [inline libc/hdr/CMakeLists.txt:24]: ah, nice, I should rebase #87017 onto this cmake change.

File: libc/include/llvm-libc-macros/math-macros.h [hunk 2]
```diff
@@ -30,12 +30,12 @@
...
-#define FP_ILOGB0 (-INT_MAX - 1)
-#define FP_LLOGB0 (-LONG_MAX - 1)
+#define FP_ILOGB0 INT_MIN
+#define FP_LLOGB0 LONG_MIN
 
 #ifdef __FP_LOGBNAN_MIN
-#define FP_ILOGBNAN (-INT_MAX - 1)
-#define FP_LLOGBNAN (-LONG_MAX - 1)
+#define FP_ILOGBNAN INT_MIN
+#define FP_LLOGBNAN LONG_MIN
```
Comment (nickdesaulniers) [inline libc/include/llvm-libc-macros/math-macros.h:38]: What's up with changing these values? Seems orthogonal to the rest of this patch?

**Review by nickdesaulniers — COMMENTED** [review]
Thanks! I think this gives me a clearer picture of what the usage of "proxy" headers might look like (#87017).

File: libc/include/llvm-libc-macros/math-macros.h (hunk 2 above)
Comment (lntue) [reply libc/include/llvm-libc-macros/math-macros.h]: Reverted the sneaked in file.

**Review by nickdesaulniers — APPROVED** [review]

**Review by michaelrj-google — APPROVED** [review]
